programs/quarter_board/main.c

```diff
@@ -0,0 +1,192 @@
+        sensors_values[i] = value;
+        values_available[i] = true;
+    } else {
+        values_available[i] = false;
+    }
+}
+
+THD_WORKING_AREA(waThread1, 64);
+THD_FUNCTION(Thread1, arg) {
+    (void)arg;
+
+    uint8_t i;
+
+    while (1) {
+        if(i_am_middle_one) {
+            values_available[0] = false;
+        }
+
+        for(i = 0; i < QUANTITY_OF_SENSORS; i++) {
+            if(values_available[i]) {
+                send_values[i] = sensors_values[i];
+                values_available[i] = false;
+            } else {
+                send_values[i] = NOT_SEEING_VALUE;
+            }
+        }
+        chThdSleepMilliseconds(100);
+    }
+    chRegSetThreadName("ir_measurement");
+    chThdSleepMilliseconds(100);
+}
+void execute_master_command(uint16_t command_id, uint8_t data_lenght, uint8_t *buff) {
+    switch(command_id) {
+        case 0xAA:  if(buff[0] == 1) led(GREEN, 1);
+                    else led(GREEN, 0);
+
+                    if(buff[1] == 1) led(RED, 1);
+                    else led(RED, 0);
+
+                    break;
+    }
+
+}
+
+uint8_t *send_data_command(uint16_t command_id, uint8_t data_lenght, uint8_t *buff) {
+    int32_t values = 0;
+    uint8_t i;
+    switch(command_id) {
+        case 0xFF: values = send_values[0] | (send_values[1] << 16);
+                   led(GREEN, 1);
+                   led(RED, 0);
+                   for(i = 0; i < data_lenght; i++) {
+                       buff[i] = ((values >> (4 * i)) & 0xf);
+                   }
+                   break; 
+    }
+    return buff;
+}
+
+THD_WORKING_AREA(waCommanderThread, 128);
+THD_FUNCTION(CommanderThread, arg) {
+    (void)arg;
+
+    thread_t * serialp;
+    MESSAGE *messagep;
+    uint16_t cmd, command_id;
+    uint8_t data_buff[MAX_DATA_CHAR_LENGHT];
+    uint8_t data_lenght, i, rw_bit;;
+    while(1) {
+        serialp = chMsgWait();
+        messagep =(MESSAGE *) chMsgGet(serialp);
+        
+        cmd = messagep -> operation;
+
+        command_id = (cmd & 0b11111111111);
+        rw_bit = (((cmd >> OPERATION_BIT_LENGHT) & 1) + 1);
+        data_lenght = (cmd >> (OPERATION_BIT_LENGHT + RW_BIT));
+        
+        if(rw_bit == PROCESS_DATA_FROM_MASTER) {
+            for(i = 0; i < data_lenght; i++) {
+                data_buff[i] = messagep -> data[i];
+            }
+            execute_master_command(command_id, data_lenght, data_buff);
+        } else {
+            send_data_command(command_id, data_lenght, data_buff);
+            for(i = 0; i < data_lenght; i++) {
+                messagep -> data[i] = data_buff[i];
+            }
+        }
+
+        chMsgRelease(serialp, (msg_t)messagep);
+    }
+    chRegSetThreadName("commander");
+    chThdSleepMilliseconds(100);
+}
+
+static const ICUConfig icucfg_1 = {
+    ICU_INPUT_ACTIVE_HIGH,
+    2000000, // 2 MHz timer
+    read_pulse, // pulse received
+    NULL,
+    NULL,
+    ICU_CHANNEL_1,
+    0
+};
+
+static const ICUConfig icucfg_2 = {

```

Comment:
Can't we do this other way, so there aren't two almost identical structures in the code?